src/context/context.rs

```diff
@@ -0,0 +1,558 @@
+
+use context::*;
+use report::example::{ExampleReport, Failure};
+
+/// Contexts are a powerful method to make your tests clear and well organized.
+/// In the long term this practice will keep tests easy to read.
+pub struct Context<T> {
+    pub(crate) info: Option<ContextInfo>,
+    pub(crate) members: Vec<ContextMember<T>>,
+    pub(crate) before_all: Vec<Box<Fn(&mut T)>>,
+    pub(crate) before_each: Vec<Box<Fn(&mut T)>>,
+    pub(crate) after_all: Vec<Box<Fn(&mut T)>>,
+    pub(crate) after_each: Vec<Box<Fn(&mut T)>>,
+}
+
+impl<T> Context<T> {
+    pub(crate) fn new(info: Option<ContextInfo>) -> Self {
+        Context {
+            info: info,
+            members: vec![],
+            before_all: vec![],
+            before_each: vec![],
+            after_all: vec![],
+            after_each: vec![],
+        }
+    }
+}
+
+unsafe impl<T> Send for Context<T>
+where
+    T: Send,
+{
+}
+unsafe impl<T> Sync for Context<T>
+where
+    T: Sync,
+{
+}
+
+impl<T> Context<T>
+where
+    T: Clone,
+{
+    /// Open and name a new context within the current context.
+    ///
+    /// Note that the order of execution is not guaranteed to match the declaration order.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # extern crate rspec;
+    /// #
+    /// # use std::io;
+    /// # use std::sync::{Arc, Mutex};
+    /// #
+    /// # use rspec::prelude::*;
+    /// #
+    /// # pub fn main() {
+    /// #     let simple = rspec::formatter::Simple::new(io::stdout());
+    /// #     let formatter = Arc::new(Mutex::new(simple));
+    /// #     let configuration = Configuration::default().parallel(false);
+    /// #     let runner = Runner::new(configuration, vec![formatter]);
+    /// #
+    /// runner.run_or_exit(rspec::suite("a test suite", (), |ctx| {
+    ///     ctx.context("opens a context labeled 'context'", |ctx| {
+    ///         // …
+    ///     });
+    /// }));
+    /// # }
+    /// ```
+    ///
+    /// Corresponding console output:
+    ///
+    /// ```no-run
+    /// running tests
+    /// Suite "a test suite":
+    ///     Context "opens a sub-context":
+    ///         …
+    /// ```
+    ///
+    /// Available aliases:
+    ///
+    /// - [`specify`](struct.Context.html#method.specify).
+    /// - [`when`](struct.Context.html#method.when).
+    pub fn context<'a, S, F>(&mut self, name: S, body: F)
+    where
+        S: Into<Option<&'a str>>,
+        F: FnOnce(&mut Context<T>) -> (),
+        T: ::std::fmt::Debug,
+    {
+        let info = name.into().map(|name| {
+            ContextInfo {
+                label: ContextLabel::Context,
+                name: name.to_owned(),
+            }
+        });
+        self.context_internal(info, body)
```

Comment:
thought: a trait would permit a cleaner separation + put the doc in another place than the logic